ndk/src/midi/safe.rs

```diff
@@ -0,0 +1,241 @@
+//! Safe bindings for [`AMidiDevice`], [`AMidiInputPort`], and [`AMidiOutputPort`]
+//!
+//! Provides implementation of `SafeMidiDeviceBox` that ensures the current thread is attached to
+//! the Java VM when being dropped, which is required by [`AMidiDevice_release`]. All other types
+//! of this module holds an [`Arc`] of `SafeMidiDeviceBox`.
+//!
+//! [`AMidiDevice`]: https://developer.android.com/ndk/reference/group/midi#amididevice
+//! [`AMidiDevice_release`]: https://developer.android.com/ndk/reference/group/midi#amididevice_release
+//! [`AMidiInputPort`]: https://developer.android.com/ndk/reference/group/midi#amidiinputport
+//! [`AMidiOutputPort`]: https://developer.android.com/ndk/reference/group/midi#amidioutputport
+
+use std::mem::{ManuallyDrop, MaybeUninit};
+use std::ops::Deref;
+use std::ptr::{self, NonNull};
+use std::sync::Arc;
+use std::{fmt, mem};
+
+use crate::media_error::Result;
+
+use super::*;
+
+/// The owner of [`MidiDevice`]. [`SafeMidiDevice`], [`SafeMidiInputPort`], and
+/// [`SafeMidiOutputPort`] holds an [`Arc<SafeMidiDeviceBox>`]. to ensure that the underlying
+/// [`MidiDevice`] is not dropped while the safe wrappers are alive.
+///
+/// [`SafeMidiDeviceBox`] also holds a pointer to the current Java VM to attach the calling thread
+/// of [`Drop::drop`] to the VM, which is required by [`ffi::AMidiDevice_release`].
```

Comment:
So it is only the `drop()` that requires to be attached to the Java VM, all other calls are safe to call elsewhere since this implements `Deref` and all `MidiDevice` functions take an immutable borrow.

You said something similar on https://github.com/rust-mobile/ndk/pull/353#discussion_r987970649, perhaps good to mention that here (as doc-comment on the `Deref`) that it is send/sync except drop?

Otherwise let's continue discussing whether we can make this the default implementation, rather than having a threadsafe and not-threadsafe variant.

Reply:
Also don't forget to have `()` on all intradoc-links to functions :wink:

Reply:
For my use, it's 100% okay to make the "safe" one default implementation; I'm just concerned about whether there's some demand for a low-level but easy-to-use wrapper for AMidi.

Reply:
> For my use, it's 100% okay to make the "safe" one default implementation; I'm just concerned about whether there's some demand for a low-level but easy-to-use wrapper for AMidi.

In my experience these are intended to be high level wrappers, and I wouldn't worry about sneaking an `Arc` in there (as long as it is clearly annotated at the places that clone/own it, so that the caller is aware that they need to drop a bunch of things) in order to make various operations lifetime-safe, which would otherwise be impossible with Rust lifetimes and borrows. Especially if this API is intended to be used in multithreaded fashion.

Then later someone may wants to have weak references instead so that they can drop their `MidiDevice` in place and make the ports dysfunct :grimacing: 

But let's again wait for @rib, I have a hunch he has a better feeling for such APIs and use-cases.

> I mentioned that MidiDevice is Send and Sync except for Drop in the module-level doc-comment; would it be better to move it to SafeMidiDeviceBox's doc-comment?

Didn't see that, only that it holds the Java VM to implement drop but missing the `, all other functions are safe to call from any thread`.

Reply:
> Didn't see that, only that it holds the Java VM to implement drop but missing the , all other functions are safe to call from any thread.

You're right. Just added that to the module-level doc-comment of `ndk::midi::safe`.

Reply:
> Also don't forget to have () on all intradoc-links to functions 😉

Fixed!